core/types/legacy_tx.go

```diff
@@ -0,0 +1,330 @@
+
+func (tx *LegacyTx) setTxFromCeloRlpList(rlplist celoTxRlpList) {
+	tx.Nonce = rlplist.Nonce
+	tx.GasPrice = rlplist.GasPrice
+	tx.Gas = rlplist.Gas
+	tx.FeeCurrency = rlplist.FeeCurrency
+	tx.GatewayFeeRecipient = rlplist.GatewayFeeRecipient
+	tx.GatewayFee = rlplist.GatewayFee
+	tx.To = rlplist.To
+	tx.Value = rlplist.Value
+	tx.Data = rlplist.Data
+	tx.V = rlplist.V
+	tx.R = rlplist.R
+	tx.S = rlplist.S
+	tx.Hash = nil // txdata.Hash is calculated and saved inside tx.Hash()
+	tx.EthCompatible = false
+}
+
+// NewTransaction creates an unsigned legacy transaction.
+// Deprecated: use NewTx instead.
+func NewTransaction(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, feeCurrency, gatewayFeeRecipient *common.Address, gatewayFee *big.Int, data []byte) *Transaction {
+	return NewTx(&LegacyTx{
+		Nonce:               nonce,
+		To:                  &to,
+		Value:               amount,
+		Gas:                 gasLimit,
+		GasPrice:            gasPrice,
+		Data:                data,
+		FeeCurrency:         feeCurrency,
+		GatewayFeeRecipient: gatewayFeeRecipient,
+		GatewayFee:          gatewayFee,
+	})
+}
+
+// NewTransaction creates an unsigned legacy transaction.
+// Deprecated: use NewTx instead.
+func NewTransactionEthCompatible(nonce uint64, to common.Address, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte) *Transaction {
+	return NewTx(&LegacyTx{
+		Nonce:         nonce,
+		To:            &to,
+		Value:         amount,
+		Gas:           gasLimit,
+		GasPrice:      gasPrice,
+		Data:          data,
+		EthCompatible: true,
+	})
+}
+
+// NewContractCreation creates an unsigned legacy transaction.
+// Deprecated: use NewTx instead.
+func NewContractCreation(nonce uint64, amount *big.Int, gasLimit uint64, gasPrice *big.Int, feeCurrency, gatewayFeeRecipient *common.Address, gatewayFee *big.Int, data []byte) *Transaction {
+	return NewTx(&LegacyTx{
+		Nonce:               nonce,
+		Value:               amount,
+		Gas:                 gasLimit,
+		GasPrice:            gasPrice,
+		Data:                data,
+		FeeCurrency:         feeCurrency,
+		GatewayFeeRecipient: gatewayFeeRecipient,
+		GatewayFee:          gatewayFee,
+	})
+}
+
+// NewContractCreation creates an unsigned legacy transaction.
+// Deprecated: use NewTx instead.
+func NewContractCreationEthCompatible(nonce uint64, amount *big.Int, gasLimit uint64, gasPrice *big.Int, data []byte) *Transaction {
+	return NewTx(&LegacyTx{
+		Nonce:         nonce,
+		Value:         amount,
+		Gas:           gasLimit,
+		GasPrice:      gasPrice,
+		Data:          data,
+		EthCompatible: true,
+	})
+}
+
+// EncodeRLP implements rlp.Encoder
+func (tx *LegacyTx) EncodeRLP(w io.Writer) error {
+	if tx.EthCompatible {
+		return rlp.Encode(w, toEthCompatibleRlpList(*tx))
+	} else {
+		return rlp.Encode(w, toCeloRlpList(*tx))
+	}
+}
+
+// DecodeRLP implements rlp.Decoder
+func (tx *LegacyTx) DecodeRLP(s *rlp.Stream) (err error) {
+	_, size, _ := s.Kind()
```

Comment:
I think you would probably need to lift the entire logic in this function into `transaction.go:DecodeRLP`. At one point I did try having the ethcompatible tx be it's own thing, but it didn't end up working. It might have been a choice I made in where I split it, but I'm also not sure what we'd get from splitting this.